engine/src/test/java/io/zeebe/engine/util/TestStreams.java

```diff
@@ -164,6 +164,10 @@ public long getLastWrittenPosition(final String name) {
     return getLogStream(name).getCommitPosition();
   }
 
+  public long getLastProcessedPosition(final String name) {
+    return getStreamProcessor(name).getLastProcessedPositionAsync().join();
+  }
```

Comment:
Nice try but nope :sweat_smile: 

The last processed position is also set **before** the side effects are called. And it doesn't work for records that are not processed (i.e. events are skipped). (see https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/main/java/io/zeebe/engine/processing/streamprocessor/ProcessingStateMachine.java#L426)

Instead, the [callbacks](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/main/java/io/zeebe/engine/processing/streamprocessor/ProcessingStateMachine.java#L456) are invoked after the event is processed completely. And, also for skipped records. Please have a look at them. We use these callback already in the [replay property test](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStateRandomizedPropertyTest.java#L46-L49).

Reply:
You're totally right, callbacks are the only way to really know that the side effects have been performed. I really should spend more time in the engine :sweat_smile: 

I've added that you can chain multiple callbacks and changed the wait method to use that. The reason for chaining is that this is kind of a requirement for the `ProcessExecutor`, and as its also used in at least one other test which does not need a callback, I figured we should just allow that instead of passing a supplier to it.

I wasn't sure if the property should be volatile, as I guess it may be set by a different thread than the one running the test - it's probably fine as we don't rely too much on its actual value and just use it as a trigger that X has happened, but if we ever do want to do something with the value then the poll may return but then when we read it it may be an older value? Am I overthinking this? :thinking:

Reply:
I think it is okay for now :sweat_smile: